File: September Challenge/Week 1/Java/Day_3/RepeatedSubstringPattern.java
// Problem : https://leetcode.com/problems/repeated-substring-pattern/
// Solution:

class Solution {
    public boolean repeatedSubstringPattern(String s) {
        int n = s.length();
        for(int len=n/2;len>0;len--){
            if(n%len==0){
                String pattern = s.substring(0,len);
                int i = len;
                int j = len+i-1;
                while(j<n){
                    String curr= s.substring(i,j+1);
                    if(!curr.equals(pattern)) 
                        break;
                    
                    i+=len;
                    j+=len;
                }
                if(i==n) return true;
            }
        }
        return false;
    }
}

File: September Challenge/Week 1/CPP/Day_3/RepeatedSubstringPattern.cpp
class Solution {
public:
    bool repeatedSubstringPattern(string s) {
        string tmp = s+s;
        /* Remove first char to avoid to match with input string which is from 0 to s.length() */
        tmp = tmp.substr(1);
        /* Remove last char to avoid to match with input string which is appended after s.length() */
        tmp = tmp.substr(0, tmp.length()-1);
        /* if the input string present then then pattern is repeated*/
        return tmp.find(s) != string::npos ? true : false;
    }
};
